truetype/truetype.go

```diff
@@ -78,6 +123,89 @@ func readTable(ttf []byte, offsetLength []byte) ([]byte, error) {
 	return ttf[offset:end], nil
 }
 
+// nameEntryInASCII converts b, which may be UTF-16 encoded, into an ACSII string.
+func nameEntryInASCII(b []byte, utf16 bool) string {
+	var buf []byte
+	if utf16 { // Equivalent to tt_name_ascii_from_utf16.
+		j := len(b)
+		if j&1 == 1 {
+			return ""
+		}
+
+		for i := 0; i < j; i += 2 {
+			el := u16(b, i)
+			if el == 0 {
+				continue
+			} else if el < 32 || el > 127 {
+				buf = append(buf, '?')
+			} else {
+				buf = append(buf, byte(el))
+			}
+		}
+	} else { // Equivalent to tt_name_ascii_from_other.
+		for _, el := range b {
+			if el == 0 {
+				continue
+			} else if el < 32 || el > 127 {
+				buf = append(buf, '?')
+			} else {
+				buf = append(buf, el)
+			}
+		}
+	}
+
+	return string(buf)
+}
+
+// continueIf returns true if every b[offset]==values[i].
+func continueIf(b []byte, offsets, values []int) bool {
+	for i, offset := range offsets {
+		if int(u16(b, offset)) != values[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+// parseSubtables wraps the commonality in Name and parseCmap.
+func parseSubtables(b []byte, name string, firstSubtableOffset, subtableSize, valueOffset, valueLengthOffset int, offsets, values []int, isU16Offset bool) (int, int, int, error) {
+	if len(b) < 4 {
+		return 0, 0, 0, FormatError(name + " too short")
+	}
+	nsubtab := int(u16(b, 2))
+	if len(b) < subtableSize*nsubtab+firstSubtableOffset {
+		return 0, 0, 0, FormatError(name + " too short")
+	}
+	offset, length, pid, x := uint32(0), uint16(0), uint16(0), firstSubtableOffset
+	for i := 0; i < nsubtab; i++ {
+		if continueIf(b[x:], offsets, values) {
+			// We read the 16-bit Platform ID and 16-bit Platform Specific ID as a single uint32.
+			// All values are big-endian.
+			pidPsid := u32(b, x)
+			// We prefer the Unicode cmap encoding. Failing to find that, we fall
+			// back onto the Microsoft cmap encoding.
+			if pidPsid == unicodeEncoding {
+				offset, length, pid = u32(b, x+valueOffset), u16(b, x+valueLengthOffset), uint16(pidPsid>>16)
```

Comment:
It might be easier if this returned the x instead of the offset and length. The isU16Offset below looks a little hacky.